### DISCUSSION TITLE
Concern about Liquibase Hub

### DISCUSSION THREAD

**Discussion by runeflobakk** [dropwizard/dropwizard#4784] [General]
Hello!
This is not as much an issue about Dropwizard, but I wanted to express a small concern I got about Liquibase yesterday, when attempting to upgrade from v3 (3.6.3) to v4 (4.3.4).

I maintain an application which uses Dropwizard, and we like to stay ahead and keep our dependencies updated. We use the, at time of writing, latest Dropwizard v2.0.21. It may also be that we override some dependencies in `dropwizard-dependencies` to later versions, if we see that they work.

One such dependency is Liquibase, which we have actually overridden to stay on version 3.6.3 now for quite a long time. Mainly because we experienced some differences in how Liquibase generates SQL from v3.7.0 which made our migrations break, and we didn't invest time in resolving it because 3.6.3 worked just fine, and it is not really being used for anything during runtime. However yesterday I decided to try to finally bring Liquibase to latest-greatest. Resolving the anomalies for the migrations was easy, so that was nothing to worry about. But when testing and running the migrations I noticed some warnings being logged, looking like this:
```
WARN  liquibase.hub.HubUpdater - Skipping auto-registration
WARNING: The changelog file specified is not registered with any Liquibase Hub project,
so the results will not be recorded in Liquibase Hub.
To register the changelog with your Hub Project run 'liquibase registerchangelog'.
Learn more at https://hub.liquibase.com
```
Which made me a bit curious of this [Liquibase Hub](https://hub.liquibase.com/), which it looks like it is something it want to contact during our database migrations. In short, it looks to be some kind of analysis tool for displaying insights into your migrations. After some digging and debugging, I think I have verified that it actually does _not_ connect to Liquibase Hub to send anything about our database schema, but it _strongly_ nudges to register an account and set up so that data about your database seems to be sent to Liquibase each time a migration is performed.

Setting the system property `liquibase.hub.mode=OFF` disables the Liquibase Hub feature. But, in our migration setup for our automated tests, we run [Liquibase.dropAll](https://github.com/liquibase/liquibase/blob/master/liquibase-core/src/main/java/liquibase/Liquibase.java#L1758-L1763), and for that I _also_ had to make sure that I use a _separate_ `Liquibase` instance where `changeLog` is explicitly set to `null`, or I would get the aforementioned warning, even though I had set the system property `liquibase.hub.mode=OFF`. It does not seem like the `DropAllCommand` in liquibase takes the system property into account.

All this makes me a bit uneasy about Liquibase, to be honest. Yes, it will seemingly not send anything as long as you have not configured an API-key to connect to their Liquibase Hub API, but it took me a bit off-guard that they are actually thinking of sending lots of data about our database to their servers, and also having it as a sort of, and a bit hidden tbh, opt-out feature. If you decide to do as the warnings advise you to, according to their FAQ they will send "Changeset body from the changelog", "Generated SQL executed by the Liquibase core app", and "Two types of logs: operation-event and change-event".

I see that Liquibase v4 is planned for the 2.1 release of Dropwizard, so I wanted to ask if you have any thoughts on this? Have you noticed this behaviour? I do not wish to be dramatic about this, am I overreacting or is this a legitimate concern? Maybe there are some insight here as well that is applicable for incorporating into Dropwizard.

Thanks for reading my rambling wall-of-text, and thanks for making Dropwizard! :)

**Comment by joschi** [comment]
@runeflobakk Thanks for bringing this up!

First of all, thanks for using Dropwizard and the `dropwizard-migrations` module. 😃

> All this makes me a bit uneasy about Liquibase, to be honest. Yes, it will seemingly not send anything as long as you have not configured an API-key to connect to their Liquibase Hub API, but it took me a bit off-guard that they are actually thinking of sending lots of data about our database to their servers, and also having it as a sort of, and a bit hidden tbh, opt-out feature. 

I agree with this sentiment and wouldn't want to ship Dropwizard 2.1.x with this feature enabled by default.

For completeness sake, there's also the [`dropwizard-flyway`](https://github.com/dropwizard/dropwizard-flyway) to use [Flyway](https://flywaydb.org/) instead of Liquibase for database migrations and it seems that Flyway will eventually also have a SaaS offering named [Flightpath](https://flywaydb.org/blog/a-first-view-of-flightpath) in the future. Let's see how they will handle the integration and whether they'll make it opt-in or opt-out.

This all being said, we'd be happy to receive a PR against the `master` or `release/2.1.x` branch to disable the Liquibase Hub feature in Dropwizard (`dropwizard-migrations`) by default and make it opt-in for interested users.


> But, in our migration setup for our automated tests, we run [Liquibase.dropAll](https://github.com/liquibase/liquibase/blob/master/liquibase-core/src/main/java/liquibase/Liquibase.java?rgh-link-date=2021-04-27T10%3A03%3A02Z#L1758-L1763), and for that I _also_ had to make sure that I use a _separate_ `Liquibase` instance where `changeLog` is explicitly set to `null`, or I would get the aforementioned warning, even though I had set the system property `liquibase.hub.mode=OFF`.

It looks like this is a bug in the `DropAllCommand` of Liquibase in which the Hub mode is only checked against the string "off" in a case-sensitive way, so that only `liquibase.hub.mode=off` would work:

https://github.com/liquibase/liquibase/blob/aa745c09ff3f0a15120b72549110e192197188e1/liquibase-core/src/main/java/liquibase/command/core/DropAllCommand.java#L124-L126
```java
if (apiKey == null || hubMode.equals("off") || !hubServiceFactory.isOnline()) {
    return null;
}
```

Maybe you could report this to the Liquibase project in [a bug report](https://github.com/liquibase/liquibase/issues/new?assignees=&labels=Status%3ADiscovery&template=bug_report.md&title=).

**Comment by zUniQueX** [comment]
@runeflobakk Do you have a minimal example for this? In the tests of the commands and in a new application with Dropwizard 2.1.0-beta.4 (which uses Liquibase 4.4.3), I don't get this warning.

The issuing line for the first warning should be this one: [HubUpdater#L476](https://github.com/liquibase/liquibase/blob/4.4.x/liquibase-core/src/main/java/liquibase/hub/HubUpdater.java#L476). However, in the standard configuration this line never gets executed, since the method returns at [HubUpdater#L407](https://github.com/liquibase/liquibase/blob/4.4.x/liquibase-core/src/main/java/liquibase/hub/HubUpdater.java#L407), because the default value for the property `allowPrompt` is false: [ConsoleUIService#L23](https://github.com/liquibase/liquibase/blob/4.4.x/liquibase-core/src/main/java/liquibase/ui/ConsoleUIService.java#L23).

**Comment by runeflobakk** [comment]
@zUniQueX I left this a bit after the "assessment" above, but I will check a branch I have here with the changes I did as soon as I find the time, and get back to you!

**Comment by zUniQueX** [comment]
@runeflobakk Any update on this?

**Comment by runeflobakk** [comment]
@zUniQueX First, sorry about the radio silence on my part. The usual excuse, things have been hectic.

I found the branch of the project I initially discovered this, removed setting `liquibase.hub.mode=off` and confirmed the warn is still emitted. I then upgraded Liquibase from 4.3.4 to 4.4.3, and the warning does indeed go away, as you describe.

So I then hope this means that efforts has been done in Liquibase to ensure that any logic related to uploading things to Liquibase Hub are not "within reach" when running Liquibase normally, unless one has actually _opted in_ to use Liquibase Hub :) 

I take it that `liquibase.hub.mode=off` should be redundant with Liquibase >= 4.4.3, or should it still explicitly turned off? If the Liquibase Hub feature is still active by default (only the warning has disappeared), I am curious about what would trigger a successful authentication against the API and consequentially uploading stuff? Do they support any "environment configured credentials", say an API-token in an environment variable on a machine of a developer who has her own account, could that suddenly be picked up by Liquibase and start to upload data about our database to their servers? Maybe I am being paranoid here, but I am not so fond of such features being opt-out, rather than opt-in.

It is also possible to programatically turn off Liquibase Hub, it seems: https://github.com/liquibase/liquibase/issues/1741#issuecomment-892501130

Thanks for giving it attention!

**Comment by zUniQueX** [comment]
@runeflobakk Thanks for the detailed answer. I just looked into their code (current `master` branch) and they seem to instantiate a `HubUpdater` anyway in the `Liquibase.update(...)` method. This instance shouldn't do anything, if the user is not registered, but I also think this should be opt-in rather than opt-out.

One solution seems to be ensuring `HubService.isOnline()` returns `false`. Then the methods of `HubUpdater` would all return before sending any data.

I just didn't fully get their configuration mechanism, so I currently can't say, if there is a default instance of `HubService` registered, which fulfills `isOnline()`. I'll take a closer look at that and will report this here.

If there is no default registered `HubService`, then a fallback one is used which directly returns at all methods instead of connecting to the hub. So in this case, we shouldn't have to care about any data send although there might be an API key anywhere.
